### DISCUSSION TITLE
Gzip response is not being decompressed automatically

### DISCUSSION THREAD

**Discussion by motishaku** [fastapi/fastapi#8571] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
from fastapi.middleware.gzip import GZipMiddleware
from fastapi import FastAPI
app = FastAPI(
    title="xyz",
    description="",
    version="1.0.0",
)

app.add_middleware(GZipMiddleware)

@app.get("/status", status_code=200, tags=["Health Check"])
async def health_check(request: Request):
    logger.info(f"health check was requested from {request.client.host}")
    return
```

### Description

Hi,

I am trying to compress the data given from our API, as it sometimes include very large JSON data sets.
i was looking into the [built-in GZIP middleware ](https://fastapi.tiangolo.com/advanced/middleware/?h=gzip#gzipmiddleware).

I am not sure if what i encounter is an issue with FastAPI or something that i'm doing wrong, because when using some other endpoints that have GZIP response (like https://postman-echo.com/gzip) the response is being decompressed automatically both from postman and the browser itself.

Also, when trying to look for it online I could not find anything, which makes me believe something i do is wrong.

when trying to access the health check endpoint, i get the following response (should be stringified null):
![image](https://user-images.githubusercontent.com/63312369/151965713-c9caa629-05ee-4d6c-abc3-77540f313386.png)

Same result is being received in the browser itself.

Am i missing something?

Thanks in advance.

### Operating System

Linux, macOS

### FastAPI Version

0.63.0

### Python Version

3.8.0

**Comment by rafsaf** [comment]
This is weird, it should be working.

What exactly server of yours sends in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding `Content-Encoding` header in **RESPONSE**? It looks for me that for some reason you overwrite this header?

It should be equal to `Content-Encoding: gzip` and then with any luck it is properly handled by postman, browser etc.

**Comment by motishaku** [comment]
hey @rafsaf, thank you for the reply.
I did not overwrite this header, and it does being sent as gzip.

Any more idea i can try?

thanks.

**Comment by motishaku** [comment]
Update:
the problem seems to be with the proxy-headers flag of uvicorn.

when i try to start my API like so:

    uvicorn.run("main:app", host="0.0.0.0", port=8999, log_level="info", reload=True, proxy_headers=True, lifespan="on")

it does not work, but doing it like that:
    uvicorn.run("main:app", host="0.0.0.0", port=8999, log_level="info", reload=True, lifespan="on")

does work.

We need to have proxy headers enabled for logging purposes behind a reverse proxy.

Any idea on what we can do?

**Comment by rafsaf** [comment]
Interesting stuff, what about other custom headers?

for example add `Trick-Or-Treat: Treat`

Does specifically proxy headers make a difference here or any non-standard headers at all, that is the only question comes to my mind.

**Comment by ghost** [comment]
I am also facing this issue, if I enable gzip middleware then the response fails to render in insomnia and chrome as well.
Any idea how to fix?

![image](https://user-images.githubusercontent.com/83651106/169515183-6f9ecab7-3b1d-417f-b994-9cfaf1ad6ac0.png)

**Comment by rafsaf** [comment]
As i remember this issue, the problem is with headers your server send. Author had proxy headers from uvicorn enabled and that is something that broke postman, chrome preview etc.

Please share exact headers you get from server using `curl`

For example for "working" example of g zipped response https://postman-echo.com/gzip 
```bash
$ rafsaf@SZCZERY:~$ curl --head https://postman-echo.com/gzip
```
Results with
```bash
HTTP/2 200 
date: Sat, 21 May 2022 05:27:05 GMT
content-type: application/json; charset=utf-8
content-length: 207
content-encoding: gzip
etag: W/"cf-Tazgf9i7aAH61KGJVvSKgCv7nPQ"
vary: Accept-Encoding
set-cookie: sails.sid=xxx
```

Then you can play around trying to change your response disabling or enabling some headers.

**Comment by JoshYuJump** [comment]
You can enhanced `APIRoute` by `get_route_handler`.
```python
class GzipRequest(Request):
    async def body(self) -> bytes:
        if not hasattr(self, "_body"):
            body = await super().body()
            if "gzip" in self.headers.getlist("Content-Encoding"):
                body = gzip.decompress(body)
            self._body = body
        return self._body


class GzipRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            request = GzipRequest(request.scope, request.receive)
            return await original_route_handler(request)

        return custom_route_handler
```

And then bind to the FastAPI app's route_class：
```python
app = FastAPI(**self.base_settings)
app.router.route_class = GzipRoute
```

ref: https://github.com/bali-framework/bali/blob/main/bali/application.py

**Comment by ghost** [comment]
@rafsaf This is the response headers I am receiving: 

I am starting the app like this ` uvicorn app.main:app --reload`

do i need to pass any other parameter?

<img width="405" alt="image" src="https://user-images.githubusercontent.com/83651106/169640234-50c188fd-d174-4859-82e5-562b99636741.png">

I am using brotli compression here but it's same with gzip also

**Comment by ghost** [comment]
> You can enhanced `APIRoute` by `get_route_handler`.
> 
> ```python
> class GzipRequest(Request):
>     async def body(self) -> bytes:
>         if not hasattr(self, "_body"):
>             body = await super().body()
>             if "gzip" in self.headers.getlist("Content-Encoding"):
>                 body = gzip.decompress(body)
>             self._body = body
>         return self._body
> 
> 
> class GzipRoute(APIRoute):
>     def get_route_handler(self) -> Callable:
>         original_route_handler = super().get_route_handler()
> 
>         async def custom_route_handler(request: Request) -> Response:
>             request = GzipRequest(request.scope, request.receive)
>             return await original_route_handler(request)
> 
>         return custom_route_handler
> ```
> 
> And then bind to the FastAPI app's route_class：
> 
> ```python
> app = FastAPI(**self.base_settings)
> app.router.route_class = GzipRoute
> ```
> 
> ref: https://github.com/bali-framework/bali/blob/main/bali/application.py

It looks like body is being decompressed before sending the response, shouldn't client handle the decompression?
If my server is sending the decompressed response then why to compress it at first place?

Could you elaborate? I might be wrong here.

**Comment by rafsaf** [comment]
Are you sure you're using GZipMiddleware as stated in docs?
`from fastapi.middleware.gzip import GZipMiddleware`
https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware

See its code in https://github.com/encode/starlette/blob/master/starlette/middleware/gzip.py#L66, FastAPI imports it directly from Starlette, in line 66 it set response header content-encoding to "gzip", and you have "br".

And by the way, it uses gzip only for bigger responses, and i don't see any reason to use custom gzip middleware if we have one smart, extensible and out of the box @JoshYuJump.

**Comment by ghost** [comment]
@rafsaf no, I was using a broti-compression compression package, was experimenting with brotli and gzip

but it is same chrome unable to render with the 3rd party library as well as the provided GZip middleware

Below is the screenshot if using the provided GZip Middleware
 
<img width="864" alt="image" src="https://user-images.githubusercontent.com/83651106/169641444-1735edf7-a89e-4a8e-a343-6ef6a8f36e37.png">

**Comment by rafsaf** [comment]
I've pasted "working" example of response headers few posts above, the only difference is with `vary` and `content-type`, first one has doubled headers but it doesn't matter but in content type header you don't send `charset`, https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type, maybe its something chrome could not stand? But it should be by default utf-8, I'm not expert in this topic.

Can you spend some time investigating it? Just try modifying headers through fastapi or proxy or whatever and find exactly where (in which step) there is a problem? It could be with uvicorn or middleware itself or fastapi or your code.

**Comment by JoshYuJump** [comment]
> Are you sure you're using GZipMiddleware as stated in docs? `from fastapi.middleware.gzip import GZipMiddleware` https://fastapi.tiangolo.com/advanced/middleware/#gzipmiddleware
> 
> See its code in https://github.com/encode/starlette/blob/master/starlette/middleware/gzip.py#L66, FastAPI imports it directly from Starlette, in line 66 it set response header content-encoding to "gzip", and you have "br".
> 
> And by the way, it uses gzip only for bigger responses, and i don't see any reason to use custom gzip middleware if we have one smart, extensible and out of the box @JoshYuJump.

@rafsaf I tried Starlette's GZipMiddleware before, but It seems can not decompress the gzip requests. It works well in gzip responses.

**Comment by ghost** [comment]
@rafsaf ok I will try to investigate, although I did try to modify the response object to add content-type with "application/json, charset=utf-8" using this example https://fastapi.tiangolo.com/advanced/response-headers/

It added another header for content-type but still chrome couldn't decompress :(

My other plan is to use nginx for compression once I deploy, can't figure out why fast api isn't working.
I had worked with django in the past with gzip compression and it was perfect.

**Comment by rafsaf** [comment]
Yup, I see.

@JoshYuJump has fair point that if it doesn't work and "something is no yes" with bundled Gzip Middleware or compression at all in FastAPI without clear reason (at least for three of us), then using custom based on build-in can be a nice alternative.

But it would be nice if we know that reason and create issue in this project or related one or document it if it is kind of expected and it actually works.

Have a nice day then! :P

**Comment by ghost** [comment]
@JoshYuJump your solution is for Requests whose body is gzip compressed, right?

But for me chrome is unable to handle gzip Response sent from uvicorn

**Comment by ghost** [comment]
Ah finally found the issue, I was using app.mount to mount the same app at the '/api' endpoint.

Like this `app.mount('/api', app)`
this might have messed up something
removing this mount point fixed the issue, thanks!
